### PR TITLE
8294924: JvmtiExport::post_exception_throw() doesn't deal well with concurrent stack processing

### PR DESCRIPTION
**Pull request by fisk**
There is a stack walk in JvmtiExport::post_exception_throw() that has safepoints in it. This trips up the stack watermark code. This patch adds a RAII object to JvmtiExport::post_exception_throw() that keeps the thread and its stack fully processed throughout the function.
Testing: tier1-7 of ZGC tests on linux x86_64 debug and manual testing of the test that failed.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8294924](https://bugs.openjdk.org/browse/JDK-8294924): JvmtiExport::post_exception_throw() doesn't deal well with concurrent stack processing

### Reviewers
 * [Patricio Chilano Mateo](https://openjdk.org/census#pchilanomate) (@pchilano - **Reviewer**) ⚠️ Review applies to [6aabc24b](https://git.openjdk.org/jdk/pull/11238/files/6aabc24bb3ac1d1763fd61d71f71881c1cc8109e)
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11238/head:pull/11238` \
`$ git checkout pull/11238`

Update a local copy of the PR: \
`$ git checkout pull/11238` \
`$ git pull https://git.openjdk.org/jdk pull/11238/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11238`

View PR using the GUI difftool: \
`$ git pr show -t 11238`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11238.diff">https://git.openjdk.org/jdk/pull/11238.diff</a>

</details>

### CODE REVIEW

**Review by pchilano — APPROVED** [review]
Looks good to me! The issue can also be easily reproduced by running a simplified variation of test events/Exception/exception01/exception01.java plus forcing a GC safepoint in that method: https://github.com/pchilano/jdk/commit/9e61eb0e0d624608d981754d3e8b6f50a6accd96

Thanks,
Patricio

File: src/hotspot/share/prims/jvmtiExport.cpp [hunk 1]
```diff
@@ -1964,6 +1965,7 @@ void JvmtiExport::post_exception_throw(JavaThread *thread, Method* method, addre
   HandleMark hm(thread);
   methodHandle mh(thread, method);
   Handle exception_handle(thread, exception);
+  KeepStackGCProcessedMark ksgcpm(thread);
```
Comment (sspitsyn) [inline src/hotspot/share/prims/jvmtiExport.cpp:1969]: Nit: It'd be nice to place a small comment about why it is needed.

**Review by sspitsyn — APPROVED** [review]
Thank you for fixing this!
Looks good.
Added a minor comment.
Thanks,
Serguei

File: src/hotspot/share/prims/jvmtiExport.cpp (hunk 1 above)
Comment (dholmes-ora) [reply src/hotspot/share/prims/jvmtiExport.cpp:1969]: +1 I've never even heard of this thing let alone understand when I would need to use it.

**Comment by fisk** [comment]
Thanks for the reviews, @dholmes-ora, @pchilano and @sspitsyn! I added a comment as requested.

**Review by dholmes-ora — APPROVED** [review]
Thanks for the added comment. I can approve this as-is but I think we have a significant problem here as it remains completely unclear when `KeepStackGCProcessedMark` is needed or how its omission would be detected. This seems extremely fragile.

**Comment by sspitsyn** [comment]
Thank you for the update. I agree with David that it is not very clear when it is needed. Approving anyway.

**Review by sspitsyn — APPROVED** [review]

**Pull request closed without merging**